### ISSUE TITLE
PATCH: important perlfunc version guards, plus outstanding sprintf TODO reminder for 5.14

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11302]
Migrated from [rt.perl.org#89758](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=89758) (status was 'resolved')

Searchable as RT89758$

**Comment by p5pRT** [comment]
### From tchrist@perl.com

This reminds users to put version guards on their neologisms\.
We're changing Perl's basic sytnax a great deal\, and users need
to understand that to use those syntactic changes will cause
weird errors if they don't put this sort of thing in\.  This 
sort of thing really should have gone in all along\.  Let's 
please continue what I have begun\.  

I have \*not\* done all the new\-falutin' sprintf stuff in
perldelta\, which I consider the release to have no business going
out without\. Don't put something in perldelta that doesn't go
elsewhere&#8203;: perldelta is the last place it should go\, not the
first\. When those broken sprintf bits \*are\* fixed\, I strongly
urge that similar warnings be included with them\.

I also fixed the order&#8203;: untie does not precede unshift\. :\(

\-\-tom

**Comment by p5pRT** [comment]
### From tchrist@perl.com

<details><summary>perldelta-version-guard.patch</summary>

```diff
--- perlfunc.pod	2011-04-25 19:40:47.000000000 -0600
+++ /tmp/perlfunc.pod	2011-05-03 11:25:49.000000000 -0600
@@ -12,7 +12,7 @@
 operators take more than one argument, while unary operators can never
 take more than one argument.  Thus, a comma terminates the argument of
 a unary operator, but merely separates the arguments of a list
-operator.  A unary operator generally provides a scalar context to its
+operator.  A unary operator generally provides scalar context to its
 argument, while a list operator may provide either scalar or list
 contexts for its arguments.  If it does both, scalar arguments 
 come first and list argument follow, and there can only ever
@@ -55,8 +55,8 @@
 C<time() + 86_400>.
 
 For functions that can be used in either a scalar or list context,
-nonabortive failure is generally indicated in a scalar context by
-returning the undefined value, and in a list context by returning the
+nonabortive failure is generally indicated in scalar context by
+returning the undefined value, and in list context by returning the
 empty list.
 
 Remember the following important rule: There is B<no rule> that relates
@@ -430,6 +430,11 @@
 the return value of C<-f $file> as an argument to another filetest
 operator, no special magic will happen.)
 
+To avoid confusing would-be users of your code with mysterious
+syntax errors, put something like this at the top of your script:
+
+    use 5.009_01;  # so filetest ops can stack
+
 =item abs VALUE
 X<abs> X<absolute>
 
@@ -1473,11 +1478,12 @@
 =item each ARRAY
 X<array, iterator>
 
-When called in list context, returns a 2-element list consisting of the key
-and value for the next element of a hash, or the index and value for the
-next element of an array, so that you can iterate over it.  When called in
-scalar context, returns only the key (not the value) in a hash, or the index
-in an array.
+When called on a hash in list context, returns a 2-element list
+consisting of the key and value for the next element of a hash.  In Perl
+5.11 and later only, it will also return the index and value for the next
+element of an array so that you can iterate over it; older Perls consider
+this a syntax error.  When called in scalar context, returns only the key
+(not the value) in a hash, or the index in an array.
 
 Hash entries are returned in an apparently random order.  The actual random
 order is subject to change in future versions of Perl, but it is
@@ -1488,14 +1494,15 @@
 
 After C<each> has returned all entries from the hash or array, the next
 call to C<each> returns the empty list in list context and C<undef> in
-scalar context.  The next call following that one restarts iteration.  Each
-hash or array has its own internal iterator, accessed by C<each>, C<keys>,
-and C<values>.  The iterator is implicitly reset when C<each> has reached
-the end as just described; it can be explicitly reset by calling C<keys> or
-C<values> on the hash or array.  If you add or delete a hash's elements
-while iterating over it, entries may be skipped or duplicated--so don't do
-that.  Exception: It is always safe to delete the item most recently
-returned by C<each()>, so the following code works properly:
+scalar context; the next call following I<that> one restarts iteration.
+Each hash or array has its own internal iterator, accessed by C<each>,
+C<keys>, and C<values>.  The iterator is implicitly reset when C<each> has
+reached the end as just described; it can be explicitly reset by calling
+C<keys> or C<values> on the hash or array.  If you add or delete a hash's
+elements while iterating over it, entries may be skipped or duplicated--so
+don't do that.  Exception: In the current implementation, it is always safe
+to delete the item most recently returned by C<each()>, so the following
+code works properly:
 
         while (($key, $value) = each %hash) {
           print $key, "\n";
@@ -1516,6 +1523,14 @@
 
     while (($key,$value) = each $hashref) { ... }
 
+To avoid confusing would-be users of your code who are running earlier
+versions of Perl with mysterious syntax errors, put this sort of thing at
+the top of your file to signal that your code will work I<only> on Perls of
+a recent vintage:
+
+    use 5.011;	# so keys/values/each work on arrays
+    use 5.014;	# so keys/values/each work on scalars (experimental)
+
 See also C<keys>, C<values> and C<sort>.
 
 =item eof FILEHANDLE
@@ -1692,7 +1707,7 @@
 in case 6.
 
 Before Perl 5.14, the assignment to C<$@> occurred before restoration 
-of localised variables, which means that for your code to run on older
+of localized variables, which means that for your code to run on older
 versions, a temporary is required if you want to mask some but not all
 errors:
 
@@ -2396,7 +2411,7 @@
 integer, in which case the result is a packed integer, which you can decode
 using C<unpack> with the C<i> (or C<I>) format.
 
-An example to test whether Nagle's algorithm is turned on on a socket:
+An example to test whether Nagle's algorithm is enabled on a socket:
 
     use Socket qw(:all);
 
@@ -2643,8 +2658,10 @@
 
 =item keys ARRAY
 
-Returns a list consisting of all the keys of the named hash, or the indices
-of an array. (In scalar context, returns the number of keys or indices.)
+Called in list context, returns a list consisting of all the keys of the
+named hash, or in Perl 5.11 or later only, the indices of an array.  Perl
+releases prior to 5.11 will produce a syntax error if you try to use an
+array argument.  In scalar context, returns the number of keys or indices.
 
 The keys of a hash are returned in an apparently random order.  The actual
 random order is subject to change in future versions of Perl, but it
@@ -2706,6 +2723,14 @@
     for (keys $hashref) { ... }
     for (keys $obj->get_arrayref) { ... }
 
+To avoid confusing would-be users of your code who are running earlier
+versions of Perl with mysterious syntax errors, put this sort of thing at
+the top of your file to signal that your code will work I<only> on Perls of
+a recent vintage:
+
+    use 5.011;	# so keys/values/each work on arrays
+    use 5.014;	# so keys/values/each work on scalars (experimental)
+
 See also C<each>, C<values> and C<sort>.
 
 =item kill SIGNAL, LIST
@@ -3710,9 +3735,9 @@
 with it.
 
 The exact semantics and interface of TYPE and ATTRS are still
-evolving.  TYPE is currently bound to the use of C<fields> pragma,
-and attributes are handled using the C<attributes> pragma, or starting
-from Perl 5.8.0 also via the C<Attribute::Handlers> module.  See
+evolving.  TYPE is currently bound to the use of the C<fields> pragma,
+and attributes are handled using the C<attributes> pragma, or, starting
+from Perl 5.8.0, also via the C<Attribute::Handlers> module.  See
 L<perlsub/"Private Variables via my()"> for details, and L<fields>,
 L<attributes>, and L<Attribute::Handlers>.
 
@@ -4111,7 +4136,11 @@
 C<< > >> or C<< < >> endianness modifiers to respectively enforce big-
 or little-endian byte-order.  These modifiers are especially useful 
 given how C<n>, C<N>, C<v> and C<V> don't cover signed integers, 
-64-bit integers, or floating-point values.
+64-bit integers, or floating-point values.  To avoid confusing
+would-be users of your code with mysterious failures, put something
+like at the top your code:
+
+    use 5.009_02;	# for pack format endianness modifiers
 
 Here are some concerns to keep in mind when using an endianness modifier:
 
@@ -4383,6 +4412,13 @@
 considered highly experimental.  The exact behaviour may change in a future
 version of Perl.
 
+To avoid confusing would-be users of your code who are running earlier
+versions of Perl with mysterious syntax errors, put this sort of thing at
+the top of your file to signal that your code will work I<only> on Perls of
+a recent vintage:
+
+    use 5.014;	# so push/pop/etc work on scalars (experimental)
+
 =item pos SCALAR
 X<pos> X<match, position>
 
@@ -4496,6 +4532,13 @@
 C<push> is considered highly experimental.  The exact behaviour may change
 in a future version of Perl.
 
+To avoid confusing would-be users of your code who are running earlier
+versions of Perl with mysterious syntax errors, put this sort of thing at
+the top of your file to signal that your code will work I<only> on Perls of
+a recent vintage:
+
+    use 5.014;	# so push/pop/etc work on scalars (experimental)
+
 =item q/STRING/
 
 =item qq/STRING/
@@ -4640,6 +4683,13 @@
     }
     closedir $dh;
 
+To avoid confusing would-be users of your code who are running earlier
+versions of Perl with mysterious failures, put this sort of thing at the
+top of your file to signal that your code will work I<only> on Perls of a
+recent vintage:
+
+    use 5.011_02; # so readdir assigned to $_ in a lone while test
+
 =item readline EXPR
 
 =item readline
@@ -5418,6 +5468,13 @@
 C<shift> is considered highly experimental.  The exact behaviour may change
 in a future version of Perl.
 
+To avoid confusing would-be users of your code who are running earlier
+versions of Perl with mysterious syntax errors, put this sort of thing at
+the top of your file to signal that your code will work I<only> on Perls of
+a recent vintage:
+
+    use 5.014;	# so push/pop/etc work on scalars (experimental)
+
 See also C<unshift>, C<push>, and C<pop>.  C<shift> and C<unshift> do the
 same thing to the left end of an array that C<pop> and C<push> do to the
 right end.
@@ -5797,6 +5854,13 @@
 C<splice> is considered highly experimental.  The exact behaviour may
 change in a future version of Perl.
 
+To avoid confusing would-be users of your code who are running earlier
+versions of Perl with mysterious syntax errors, put this sort of thing at
+the top of your file to signal that your code will work I<only> on Perls of
+a recent vintage:
+
+    use 5.014;	# so push/pop/etc work on scalars (experimental)
+
 =item split /PATTERN/,EXPR,LIMIT
 X<split>
 
@@ -6260,11 +6324,14 @@
 
 Sets and returns the random number seed for the C<rand> operator.
 
-The point of the function is to "seed" the C<rand> function so that
-C<rand> can produce a different sequence each time you run your
-program.  When called with a parameter, C<srand> uses that for the seed;
-otherwise it (semi-)randomly chooses a seed.  In either case, starting with
-Perl 5.14, it returns the seed.
+The point of the function is to "seed" the C<rand> function so that C<rand>
+can produce a different sequence each time you run your program.  When
+called with a parameter, C<srand> uses that for the seed; otherwise it
+(semi-)randomly chooses a seed.  In either case, starting with Perl 5.14,
+it returns the seed.  To signal that your code will work I<only> on Perls
+of a recent vintage:
+
+    use 5.014;	# so srand returns the seed
 
 If C<srand()> is not called explicitly, it is called implicitly without a
 parameter at the first use of the C<rand> operator.  However, this was not true
@@ -7257,12 +7324,6 @@
 
 See L</pack> for more examples and notes.
 
-=item untie VARIABLE
-X<untie>
-
-Breaks the binding between a variable and a package.  (See C<tie>.)
-Has no effect if the variable is not tied.
-
 =item unshift ARRAY,LIST
 X<unshift>
 
@@ -7281,6 +7342,19 @@
 C<unshift> is considered highly experimental.  The exact behaviour may
 change in a future version of Perl.
 
+To avoid confusing would-be users of your code who are running earlier
+versions of Perl with mysterious syntax errors, put this sort of thing at
+the top of your file to signal that your code will work I<only> on Perls of
+a recent vintage:
+
+    use 5.014;	# so push/pop/etc work on scalars (experimental)
+
+=item untie VARIABLE
+X<untie>
+
+Breaks the binding between a variable and a package.  (See C<tie>.)
+Has no effect if the variable is not tied.
+
 =item use Module VERSION LIST
 X<use> X<module> X<import>
 
@@ -7451,25 +7525,25 @@
 
 =item values ARRAY
 
-Returns a list consisting of all the values of the named hash, or the values
-of an array. (In a scalar context, returns the number of values.)
-
-The values are returned in an apparently random order.  The actual
-random order is subject to change in future versions of Perl, but it
-is guaranteed to be the same order as either the C<keys> or C<each>
-function would produce on the same (unmodified) hash.  Since Perl
-5.8.1 the ordering is different even between different runs of Perl
-for security reasons (see L<perlsec/"Algorithmic Complexity Attacks">).
+In list context, returns a list consisting of all the values of the named
+hash.  In Perl 5.11 or later only, will also return a list of the values of
+an array; prior to that release, attempting to use an array argument will
+produce a syntax error.  In scalar context, returns the number of values.
+
+When called on a hash, the values are returned in an apparently random
+order.  The actual random order is subject to change in future versions of
+Perl, but it is guaranteed to be the same order as either the C<keys> or
+C<each> function would produce on the same (unmodified) hash.  Since Perl
+5.8.1 the ordering is different even between different runs of Perl for
+security reasons (see L<perlsec/"Algorithmic Complexity Attacks">).
 
 As a side effect, calling values() resets the HASH or ARRAY's internal
-iterator,
-see L</each>. (In particular, calling values() in void context resets
-the iterator with no other overhead. Apart from resetting the iterator,
-C<values @array> in list context is the same as plain C<@array>.
-We recommend that you use void context C<keys @array> for this, but reasoned
-that it taking C<values @array> out would require more documentation than
-leaving it in.)
-
+iterator, see L</each>. (In particular, calling values() in void context
+resets the iterator with no other overhead. Apart from resetting the
+iterator, C<values @array> in list context is the same as plain C<@array>.
+(We recommend that you use void context C<keys @array> for this, but
+reasoned that taking C<values @array> out would require more
+documentation than leaving it in.)
 
 Note that the values are not copied, which means modifying them will
 modify the contents of the hash:
@@ -7485,6 +7559,14 @@
     for (values $hashref) { ... }
     for (values $obj->get_arrayref) { ... }
 
+To avoid confusing would-be users of your code who are running earlier
+versions of Perl with mysterious syntax errors, put this sort of thing at
+the top of your file to signal that your code will work I<only> on Perls of
+a recent vintage:
+
+    use 5.011;	# so keys/values/each work on arrays
+    use 5.014;	# so keys/values/each work on scalars (experimental)
+
 See also C<keys>, C<each>, and C<sort>.
 
 =item vec EXPR,OFFSET,BITS

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue May 03 10&#8203;:33&#8203;:46 2011\, tom christiansen wrote&#8203;:
> This reminds users to put version guards on their neologisms\.
> We're changing Perl's basic sytnax a great deal\, and users need
> to understand that to use those syntactic changes will cause
> weird errors if they don't put this sort of thing in\.  This 
> sort of thing really should have gone in all along\.  Let's 
> please continue what I have begun\.  

\*Please\* don’t use dev version numbers\. Dev versions are an artefact of
the developement process\. Using them can lead to problems\. For instance\,
if we require 5\.13\.7 for keys $scalar\, that ignores the fact that 5\.14
has a different implementation\.

If we require 5\.13\.9 for ‘use re "/a"’\, \*that\* ignores the fact that the
argument to ‘use re’ is parsed differently in 5\.13\.10\.

There are so many slight differences between versions or
half\-implemented features in this or that dev version that it very
quickly stops making sense\.

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @cpansprout

On Sun May 22 13&#8203;:20&#8203;:46 2011\, sprout wrote&#8203;:
> On Tue May 03 10&#8203;:33&#8203;:46 2011\, tom christiansen wrote&#8203;:
> > This reminds users to put version guards on their neologisms\.
> > We're changing Perl's basic sytnax a great deal\, and users need
> > to understand that to use those syntactic changes will cause
> > weird errors if they don't put this sort of thing in\.  This 
> > sort of thing really should have gone in all along\.  Let's 
> > please continue what I have begun\.  
> 
> \*Please\* don’t use dev version numbers\. Dev versions are an artefact of
> the developement process\. Using them can lead to problems\. For instance\,
> if we require 5\.13\.7 for keys $scalar\, that ignores the fact that 5\.14
> has a different implementation\.
> 
> If we require 5\.13\.9 for ‘use re "/a"’\, \*that\* ignores the fact that the
> argument to ‘use re’ is parsed differently in 5\.13\.10\.
> 
> There are so many slight differences between versions or
> half\-implemented features in this or that dev version that it very
> quickly stops making sense\.

I’ve changed the numbers and applied your patch as 3d21943e\.  Thank you\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Wed Jan 04 23&#8203;:55&#8203;:37 2012\, sprout wrote&#8203;:
> On Sun May 22 13&#8203;:20&#8203;:46 2011\, sprout wrote&#8203;:
> > On Tue May 03 10&#8203;:33&#8203;:46 2011\, tom christiansen wrote&#8203;:
> > > This reminds users to put version guards on their neologisms\.
> > > We're changing Perl's basic sytnax a great deal\, and users need
> > > to understand that to use those syntactic changes will cause
> > > weird errors if they don't put this sort of thing in\.  This 
> > > sort of thing really should have gone in all along\.  Let's 
> > > please continue what I have begun\.  
> > 
> > \*Please\* don’t use dev version numbers\. Dev versions are an artefact of
> > the developement process\. Using them can lead to problems\. For instance\,
> > if we require 5\.13\.7 for keys $scalar\, that ignores the fact that 5\.14
> > has a different implementation\.
> > 
> > If we require 5\.13\.9 for ‘use re "/a"’\, \*that\* ignores the fact that the
> > argument to ‘use re’ is parsed differently in 5\.13\.10\.
> > 
> > There are so many slight differences between versions or
> > half\-implemented features in this or that dev version that it very
> > quickly stops making sense\.
> 
> I’ve changed the numbers and applied your patch as 3d21943e\.  Thank you\.
> 

Sorry\, I meant bade7fb\.

\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue May 03 10&#8203;:33&#8203;:46 2011\, tom christiansen wrote&#8203;:
> This reminds users to put version guards on their neologisms\.
> We're changing Perl's basic sytnax a great deal\, and users need
> to understand that to use those syntactic changes will cause
> weird errors if they don't put this sort of thing in\.  This 
> sort of thing really should have gone in all along\.  Let's 
> please continue what I have begun\.  
> 
> I have \*not\* done all the new\-falutin' sprintf stuff in
> perldelta\, which I consider the release to have no business going
> out without\. Don't put something in perldelta that doesn't go
> elsewhere&#8203;: perldelta is the last place it should go\, not the
> first\. When those broken sprintf bits \*are\* fixed\, I strongly
> urge that similar warnings be included with them\.

And \*that\* was fixed in 3d21943e\.

> 
> I also fixed the order&#8203;: untie does not precede unshift\. :\(
> 
> \-\-tom
> 


\-\- 

Father Chrysostomos

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
